climada/engine/uncertainty.py

```diff
@@ -0,0 +1,592 @@
+
+    def plot_distr(self):
+        """
+        Plot the distributions of the parameters of the uncertainty variable.
+
+        Returns
+        -------
+        fig, ax: matplotlib.pyplot.fig, matplotlib.pyplot.ax
+            The figure and axis handle of the plot.
+
+        """
+        nplots = len(self.distr_dict)
+        nrows, ncols = int(nplots / 3) + 1, min(nplots, 3)
+        fig, ax = plt.subplots(nrows=nrows, ncols=ncols, figsize=(20, 16))
+        for ax, (param_name, distr) in zip(ax, self.distr_dict.items()):
+            x = np.linspace(distr.ppf(0.001), distr.ppf(0.999), 100)
+            ax.plot(x, distr.pdf(x), label=param_name)
+            ax.legend()
+        return fig, ax
+
+    def eval_unc_var(self, kwargs):
+        """
+        Evaluate the uncertainty variable.
+
+        Parameters
+        ----------
+        kwargs :
+            These parameters will be passed to self.unc_var.
+            They must be the input parameters of the uncertainty variable .
+
+        Returns
+        -------
+
+            Evaluated uncertainty variable
+
+        """
+        return self.unc_var(**kwargs)
+
+
+class UncSensitivity():
+    """
+    Sensitivity analysis class
+
+    This class performs sensitivity analysis on the outputs of a
+    climada.engine.impact.Impact() or climada.engine.costbenefit.CostBenefit()
+    object.
+
+    """
+
+    def __init__(self, exp_unc, impf_unc, haz_unc, pool=None):
+        """Initialize UncSensitivite
+
+        Parameters
+        ----------
+        exp_unc : climada.engine.uncertainty.UncVar or climada.entity.Exposure
+            Exposure uncertainty variable or Exposure
+        impf_unc : climada.engine.uncertainty.UncVar or climada.entity.ImpactFuncSet
+            Impactfunction uncertainty variable or Impact function
+        haz_unc : climada.engine.uncertainty.UncVar or climada.hazard.Hazard
+            Hazard uncertainty variable or Hazard
+        pool : pathos.pools.ProcessPool
+            Pool of CPUs for parralel computations. Default is None.
+
+        Returns
+        -------
+        None.
+
+        """
+
+        if pool:
+            self.pool = pool
+            LOGGER.info('Using %s CPUs.', self.pool.ncpus)
+        else:
+            self.pool = None
+
+        if isinstance(exp_unc, Exposures):
+            self.exp = UncVar(unc_var=lambda: exp_unc, distr_dict={})
+        else:
+            self.exp = exp_unc
+
+        if isinstance(impf_unc, ImpactFuncSet):
+            self.impf = UncVar(unc_var=lambda: impf_unc, distr_dict={})
+        else:
+            self.impf = impf_unc
+
+        if isinstance(haz_unc, Hazard):
+            self.haz = UncVar(unc_var=lambda: haz_unc, distr_dict={})
+        else:
+            self.haz = haz_unc
+
+        self.params = pd.DataFrame()
+        self.problem = {}
+        self.aai_freq = pd.DataFrame()
+        self.eai_exp = pd.DataFrame()
+        self.at_event = pd.DataFrame()
+
+
+    def plot_uncertainty(self):
```

Comment:
would be cool to have something that produces a plot like the one at page two:
https://safe-insurance.uk/EGU2020_display_Noacco_et_al_13_04_20.pdf

Reply:
Yes indeed. I would suggest to do that once the Module has reached some maturity in its content/form. Maybe you could then do that :D ?

Reply:
Indeed, we could also come up with a list of stuff and distribute work - although I think you already coded the core and more, already..

Reply:
Well, here is my plan before coming to some of the graphics:

- Separate the class into an Uncertainty class, a Sensitivity sub-class and a Robustness sub-class.
- Make it possible to use different kind of sampling and different kinds of sensitivity analysis (not only Sobol)
- Make everything as smooth/fast as possible